services/motion/builtin/moveonglobe.go

```diff
@@ -0,0 +1,312 @@
+package builtin
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	geo "github.com/kellydunn/golang-geo"
+	"github.com/pkg/errors"
+	goutils "go.viam.com/utils"
+
+	"go.viam.com/rdk/components/base"
+	"go.viam.com/rdk/components/base/kinematicbase"
+	"go.viam.com/rdk/components/movementsensor"
+	"go.viam.com/rdk/motionplan"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/services/motion"
+	"go.viam.com/rdk/spatialmath"
+)
+
+type execManager struct {
+	plan                   motionplan.Plan
+	successChan            chan bool
+	positionPollingErrChan chan error
+	obsticlePollingErrChan chan error
+	executionErrChan       chan error
+	replanChan             chan bool
+	cancelCtx              context.Context
+	cancelFn               context.CancelFunc
+	backgroundWorkers      *sync.WaitGroup
+}
+
+func plan(
+	ctx context.Context,
+	planRequest *motionplan.PlanRequest,
+	kb kinematicbase.KinematicBase,
+	componentName string,
+) (motionplan.Plan, error) {
+	inputs, err := kb.CurrentInputs(ctx)
+	if err != nil {
+		return make(motionplan.Plan, 0), err
+	}
+	// TODO: this is really hacky and we should figure out a better place to store this information
+	if len(kb.Kinematics().DoF()) == 2 {
+		inputs = inputs[:2]
+	}
+	planRequest.StartConfiguration = map[string][]referenceframe.Input{componentName: inputs}
+
+	return motionplan.PlanMotion(ctx, planRequest)
+}
+
+func startPollingForReplan(
+	ctx context.Context,
+	period time.Duration,
+	errChan chan error,
+	replanChan chan bool,
+	doneFn func(),
+	fn func(context.Context,
+	) (bool, error),
+) {
+	goutils.ManagedGo(func() {
+		ticker := time.NewTicker(period)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				replan, err := fn(ctx)
```

Comment:
The big change here is that the callback function `fn` isn't responsible for writing to the success channel. Instead it just returns a boolean result which writes to the replan channel.  

Additional changes include:
Making all channels buffered with capacity 1. This makes the first channel write nonblocking & ensures that the replanning goroutines won't block waiting for the planning go routine to consume their messages. 

Also, this change ensures that only one errChan / replanChan is written to per spawned go routine, which eliminates the bug where the replan detecting go routines block on writing to an unbuffered channel twice and blocking forever.